---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# EncircleR

The goal of EncircleR is to automate and streamline circRNA analysis, by providing a graphical interface for selecting and downloading reference files, Perform read trimming and read mapping, and finally to perform circRNA detection by exploring backsplice junction contents.

# Installation and startup

## Dependencies
The package requires the following external software to be installed

 * STAR aligner (https://github.com/alexdobin/STAR), Alignment and chimeric read detection
 * fastp (https://github.com/OpenGene/fastp), Read trimming and quality parametrics
 * R (https://cran.rstudio.com/), Statistical programming language for running EncircleR
 * Rstudio (https://rstudio.com/products/rstudio/download/#download), Graphical R interface which simplifies executing and openning EncircleR

#### Locate the binary execution files
In order to make EncircleR call the STAR aligner and fastp software, you must provide the absolute path to their binary execution files.

For STAR aligner on linux, this can be located in:
```{bash, eval=FALSE}
/path/to/STAR/bin/Linux_x86_64/STAR
```

For fastp (when installed through Bioconda (Miniconda3)):
```{bash, eval=FALSE}
/path/to/bioconda/bin/fastp
```

### R packages
After Rstudio has been successfully installed a few R packages must be installed, to make this process easier, open Rstudio and follow the steps below.

Most official R package dependencies should be installed during installation of EncircleR, however there are a few packages that needs to be installed manually first:
```{r CRAN dep, eval=FALSE}
install.packages(c("devtools", "BiocManager"))
```

Once BiocManager has been installed, the following Bioconductor packages will have to be installed as well.
```{r BioC deps, eval=FALSE}
BiocManager::install(c("AnnotationHub", "plyranges"))
```


The circRNA detection algorithm (circulaR) used by EncircleR are installed via GitHub:
``` r
devtools::install_github("https://github.com/KasperThystrup/circulaR")
```

Finally, you can then install EncircleR via Github:
``` r
devtools::install_github("https://github.com/KasperThystrup/EncircleR")
```
### How to run

After EncricleR has been successfully installed, it can be executed by opening RStudio and execute:

```{r example, eval=FALSE}
EncircleR::run_app()
```

### Metadata

In order to perform the analysis, a metadata sheet must be generated.
This metadtasheet is a tsv file, which contains the following columns:

* A sample name column (It will be used for naming output files, so make it simple - No [ ,.*|] etc.)
* A file path column (to reduce risk of errors, make sure to write the full file path - e.g. /home/Me/RNAdata/Sample1/fastq/sample01_1.fq.gz)
* A read mate column (used to distinguishes read mates 1 and 2, this means that paired end reads contains two rows for each sample, the number must be an integer)



## If package will not execute with above code

Clone the repository and open `EncircleR.Rproj` with Rstudio, first enter:
 File > New project > Version control > Git
 

Next: Try to run the entire script in `dev/run_dev.R`, install any packages that is missing.

The process is stepwise, where options by default are hidden, until the appropriate steps have been taken. Note that STAR_idx takes ALOT of RAM and FASTP and STAR alignment takes a lot of cpu time as well.

The app requires gz compressed (or uncompressed) fastq files, a metadata file with the following columns:

1. Sample name
2. Full filepath
3. Read Mate

The app currently copies (option to move instead) fastq files to a experimental directory (default set in `R/utils_DEFAULTS.R`
fastp and STAR is run on each sample, and relocated systematically
Finally, the circulaR algorithm is used to run circRNA analysis (circulaR must be installed, but have not yet been uplaoded to github)
